apache2/msc_multipart.c

```diff
@@ -841,121 +841,129 @@ int multipart_init(modsec_rec *msr, char **error_msg) {
-                if ((seen_semicolon == 0) && (*p == ';')) {
-                    seen_semicolon = 1; /* It is OK to have one semicolon. */
-                } else {
-                    msr->mpd->flag_error = 1;
-                    *error_msg = apr_psprintf(msr->mp, "Multipart: Invalid boundary in C-T (malformed).");
-                    return -1;
-                }
-            }
-        }
-
-        /* Have we seen the semicolon in the header? */
-        if (seen_semicolon == 0) {
-            msr->mpd->flag_missing_semicolon = 1;
-        }
+    if (msr->mpd->boundary == NULL) { /* Could not find boundary in the C-T header. */
+        msr->mpd->flag_error = 1;
 
-        b = strchr(msr->mpd->boundary + 8, '=');
-        if (b == NULL) {
-            msr->mpd->flag_error = 1;
-            *error_msg = apr_psprintf(msr->mp, "Multipart: Invalid boundary in C-T (malformed).");
+        /* Test for case-insensitive boundary. Allowed by the RFC but highly unusual. */
+        if (multipart_count_boundary_params(msr->mp, msr->request_content_type) > 0) {
+            *error_msg = apr_psprintf(msr->mp, "Multipart: Invalid boundary in C-T (case sensitivity).");
             return -1;
         }
 
-        /* Check parameter name ends well. */
-        if (b != (msr->mpd->boundary + 8)) {
-            /* Check all characters between the end of the boundary
-             * and the = character.
-             */
-            for (p = msr->mpd->boundary + 8; p < b; p++) {
-                if (isspace(*p)) {
-                    /* Flag for whitespace after parameter name. */
-                    msr->mpd->flag_boundary_whitespace = 1;
-                } else {
-                    msr->mpd->flag_error = 1;
-                    *error_msg = apr_psprintf(msr->mp, "Multipart: Invalid boundary in C-T (parameter name).");
-                    return -1;
-                }
-            }
+        *error_msg = apr_psprintf(msr->mp, "Multipart: Boundary not found in C-T.");
+        return -1;
+    }
+
+    /* Check for extra characters before the boundary. */
+    int seen_semicolon = 0;
+    for (char *p = (char *)(msr->request_content_type + 19); p < msr->mpd->boundary; p++) {
+        if (*p == ';') {
+            seen_semicolon = 1;
         }
+    }
 
-        b++; /* Go over the = character. */
-        len = strlen(b);
+    /* Have we seen the semicolon in the header? */
+    if (seen_semicolon == 0) {
+        msr->mpd->flag_missing_semicolon = 1;
+    }
 
-        /* Flag for whitespace before parameter value. */
-        if (isspace(*b)) {
-            msr->mpd->flag_boundary_whitespace = 1;
+    /* Find the end of boundary - more attributes such as 'charset' may follow it */
+    char *boundary_end = NULL;
+    int trailing_whitespaces = 0;
+    for (boundary_end = msr->mpd->boundary; *boundary_end != '\0'; ++boundary_end) {
+        if (*boundary_end == ';') {
+            break;
+        } else if (isspace(*boundary_end)) {
```

Comment:
could this accidentally consume the \r\n after the header? Or is the header null terminated here? (I believe it's not in nginx in general, but there they have ngx_str_t, so maybe this is a null terminated copy)

Reply:
This code is in Apache so all headers are null-terminated (Nginx wrapping code takes care of that). The old version of this code also relied on the string to be null-terminated by calling routines like `strlen()` so I deemed this to be safe.